### PR TITLE
Check the pod labels before namespace deletion

### PR DESCRIPTION
**Pull request by tomahawk28**
Toward https://github.com/giantswarm/giantswarm/issues/16353

### CODE REVIEW

File: service/controller/resource/namespace/current.go [hunk 1]
```diff
@@ -71,7 +74,12 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
 	// remove the namespace as usual.
 	if key.IsDeleted(&customObject) {
 		n := key.ClusterNamespace(customObject)
-		list, err := r.k8sClient.CoreV1().Pods(n).List(ctx, metav1.ListOptions{})
+
+		lo := metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s=%s", label.ManagedBy, project.Name()),
+		}
+
+		list, err := r.k8sClient.CoreV1().Pods(n).List(ctx, lo)
```
Comment (tomahawk28) [inline service/controller/resource/namespace/current.go:82]: After this, we don't count app-operator per cluster namespace anymore.

Comment (rossf7) [reply service/controller/resource/namespace/current.go:82]: I think we set the managed-by label on the deployment but not in the template. Can you check the pods have this label?

Comment (tomahawk28) [reply service/controller/resource/namespace/current.go:82]: You're right. It's on the deployment side; 
```
 k get -n 60tzw deploy worker-zsa9a -o yaml | yq read - "metadata.labels"
app: worker
cluster: 60tzw
customer: giantswarm
giantswarm.io/cluster: 60tzw
giantswarm.io/managed-by: kvm-operator
giantswarm.io/organization: giantswarm
node: zsa9a
```

```
 k get po -n 60tzw worker-zsa9a-785788964c-qt5ht -o yaml | yq read - "metadata.labels"
app: worker
cluster: 60tzw
customer: giantswarm
giantswarm.io/cluster: 60tzw
giantswarm.io/organization: giantswarm
kvm-operator.giantswarm.io/pod-watcher: kvm-operator
node: zsa9a
pod-template-hash: 785788964c
```

I will fix this, sorry!

File: service/controller/resource/namespace/current.go [hunk 2]
```diff
@@ -71,12 +74,17 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
...
+			LabelSelector: fmt.Sprintf("%s=%s", label.ManagedBy, project.Name()),
+		}
+
+		list, err := r.k8sClient.AppsV1().Deployments(n).List(ctx, lo)
 		if err != nil {
 			return nil, microerror.Mask(err)
 		}
 		if len(list.Items) != 0 {
-			r.logger.Debugf(ctx, "cannot finish deletion of namespace due to existing pods")
+			r.logger.Debugf(ctx, "cannot finish deletion of namespace due to existing deployments")
```
Comment (tomahawk28) [inline service/controller/resource/namespace/current.go:87]: Changing the log to `deployments`  as well.

File: CHANGELOG.md [hunk 3]
```diff
@@ -20,6 +20,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
+- Checking deployments in the cluster namespace using labels.
```
Comment (rossf7) [inline CHANGELOG.md:23]:
```suggestion
- Use managed-by label to check node deployments are deleted before cluster namespace.
```

I think we should mention this is using the managed-by label. WDYT?

File: service/controller/resource/namespace/current.go (hunk 2 above)
Comment (rossf7) [reply service/controller/resource/namespace/current.go:87]: We drain the pods so does that mean the deployment is not deleted till after the pods?

If so then I think this is fine.

**Review by rossf7 — COMMENTED** [review]
Looks good just a query on the deletion logic.

File: CHANGELOG.md (hunk 3 above)
Comment (tomahawk28) [reply CHANGELOG.md:23]: Looks better, Thanks.

File: service/controller/resource/namespace/current.go (hunk 2 above)
Comment (tomahawk28) [reply service/controller/resource/namespace/current.go:87]: Yes, exactly. kvm-operator will be deleting the deployment in the deletion phase. So it's fine.  the cluster deletion phase. https://github.com/giantswarm/kvm-operator/blob/682a07b57f78d1742c27c2c42a858b7eefd77e51/service/controller/resource/deployment/delete.go#L28-L36

Comment (rossf7) [reply service/controller/resource/namespace/current.go:87]: OK cool thanks for confirming.

**Review by rossf7 — APPROVED** [review]
LGTM

**Comment by tomahawk28** [comment]
@giantswarm/team-rocket Hi! Can anyone take a look into this, please? this is a quite simple change.

**Comment by tfussell** [comment]
For my curiosity, when would there be another non-node deployment in the MC cluster namespace? Isn't the WC `app-operator` installed in the `giantswarm` namespace? Or is this just preparation for a future change?

File: service/controller/resource/namespace/current.go [hunk 4]
```diff
@@ -71,12 +74,17 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
 	// remove the namespace as usual.
 	if key.IsDeleted(&customObject) {
 		n := key.ClusterNamespace(customObject)
-		list, err := r.k8sClient.CoreV1().Pods(n).List(ctx, metav1.ListOptions{})
+
+		lo := metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s=%s", label.ManagedBy, project.Name()),
```
Comment (tfussell) [inline service/controller/resource/namespace/current.go:79]:
```suggestion
			LabelSelector: fmt.Sprintf("%s=%s", label.ManagedBy, key.OperatorName),
```
`project.Name()` and `key.OperatorName` should both be `kvm-operator`, but this just makes sure that we're following the deployment label definition here https://github.com/giantswarm/kvm-operator/blob/v3.14.2/service/controller/resource/deployment/master_deployment.go#L93.

**Review by tfussell — APPROVED** [review]
LGTM. I'll go ahead and approve. Up to you if you want to accept my suggested change.

File: service/controller/resource/namespace/current.go (hunk 4 above)
Comment (tomahawk28) [reply service/controller/resource/namespace/current.go:79]: Thanks for pointing it out!

**Pull request merged**